### PR TITLE
added layers to case

### PR DESCRIPTION
**Pull request by Blaussh**

### CODE REVIEW

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts [hunk 1]
```diff
@@ -40,7 +44,10 @@ export class SelectCaseAppEffects {
 		.mergeMap(({ payload }: SelectCaseAction) => this.selectCaseActions(payload));
 
 	constructor(protected actions$: Actions,
-				protected store$: Store<IAppState>) {
+				protected store$: Store<IAppState>,
+				protected dataLayersService: DataLayersService
+	) {
+		this.dataLayersService.getAllLayersInATree().subscribe((layersContainers) => this._allLayrs = layersContainers);
```
Comment (liorze) [inline src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts]: Unsubcribe

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts [hunk 2]
```diff
@@ -24,9 +25,12 @@ import { IAppState } from '@ansyn/ansyn/app-effects/app.effects.module';
 import { UpdateOverlaysManualProcessArgs } from '@ansyn/menu-items/tools/actions/tools.actions';
 import { UpdateFacetsAction } from '@ansyn/menu-items/filters/actions/filters.actions';
 import { CasesService } from '@ansyn/menu-items/cases/services/cases.service';
+import { LayersContainer } from '@ansyn/menu-items/layers-manager/models/layers.model';
+import { DataLayersService } from '@ansyn/menu-items/layers-manager/services/data-layers.service';
 
 @Injectable()
 export class SelectCaseAppEffects {
+	_allLayrs: LayersContainer[] = [];
```
Comment (liorze) [inline src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts]: nit: typo

File: src/app/@ansyn/plugins/openlayers/open-layers-map/openlayers-map/openlayers-map.ts [hunk 3]
```diff
@@ -59,8 +58,10 @@ export class OpenLayersMap extends IMap<OLMap> {
 		}
 
 		const layersArray: any[] = group.getLayers().getArray();
-		let removeLayer = layersArray.indexOf(layersArray.filter(l => l.id === layer.id));
-		group.getLayers().getArray().splice(removeLayer, 1);
+		let removeLayer = layersArray.indexOf(layersArray.find(l => l.id === layer.id));
```
Comment (liorze) [inline src/app/@ansyn/plugins/openlayers/open-layers-map/openlayers-map/openlayers-map.ts]: Use filter instead.

File: src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.ts [hunk 4]
```diff
@@ -27,13 +28,16 @@ export class LayersEffects {
...
 	 * @dependencies layers
-	 * @action UnselectLayerAction?, LayerCollectionLoadedAction?, SelectLayerAction?, ErrorLoadingLayersAction?
+	 * @action LayerCollectionLoadedAction?, ErrorLoadingLayersAction?
 	 */
 	@Effect()
 	beginLayerTreeLoad$: Observable<LayersActions> = this.actions$
 		.ofType(LayersActionTypes.BEGIN_LAYER_COLLECTION_LOAD)
 		.mergeMap(() => this.dataLayersService.getAllLayersInATree())
-		.map((layersContainer: LayersContainer[]) => new LayerCollectionLoadedAction(layersContainer))
+		.withLatestFrom(this.store.select(selectSelectedLayersIds))
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.ts]: Move to layers.app.effects

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts (hunk 1 above)
Comment (Blaussh) [reply src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts]: 👎

File: src/app/@ansyn/plugins/openlayers/open-layers-map/openlayers-map/openlayers-map.ts (hunk 3 above)
Comment (Blaussh) [reply src/app/@ansyn/plugins/openlayers/open-layers-map/openlayers-map/openlayers-map.ts]: 👍

File: src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.spec.ts [hunk 5]
```diff
@@ -10,38 +10,35 @@ describe('LayersReducer', () => {
...
-			name: 'complexLayers',
 			id: 'complexLayersId',
-			isChecked: false
+			name: 'complexLayers',
+			type: LayerType.complex,
+			dataLayerContainers: [],
+			creationTime: new Date()
 		};
 
 		let layers: Layer[] = [staticLayer, dynamicLayer, complexLayer];
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.spec.ts]: Why do you need this variable?

File: src/app/@ansyn/menu-items/layers-manager/models/layers.model.ts [hunk 6]
```diff
@@ -1,22 +1,15 @@
...
 	name: string;
-	id: string;
-	isChecked: boolean;
-}
-
-export interface LayersContainer extends Entity {
-	id: string;
-	name: string;
 	type: LayerType;
 	dataLayerContainers: any[];
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/models/layers.model.ts]: Is it a leftover?

File: src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.spec.ts [hunk 7]
```diff
@@ -58,41 +59,39 @@ describe('LayersEffects', () => {
...
-			name: 'complexLayers',
 			id: 'complexLayersId',
-			isChecked: false,
+			name: 'complexLayers',
+			type: LayerType.complex,
+			dataLayerContainers: [],
+			creationTime: new Date()
 		};
 
 		let layers: Layer[] = [staticLayer, dynamicLayer, complexLayer];
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.spec.ts]: Unused

File: src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.spec.ts [hunk 8]
```diff
@@ -58,41 +59,39 @@ describe('LayersEffects', () => {
 	it('beginLayerTreeLoad$ should call dataLayersService.getAllLayersInATree with case id from state, and return LayerCollectionLoadedAction', () => {
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.spec.ts]: Test description should be changed

File: src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.spec.ts [hunk 9]
```diff
@@ -10,38 +10,35 @@ describe('LayersReducer', () => {
 	it('LAYER_COLLECTION_LOADED action should add the new tree to the state', () => {
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.spec.ts]: It's not a tree anymore..

File: src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts [hunk 10]
```diff
@@ -17,9 +17,9 @@ import { groupBy } from 'lodash';
 export class LayersManagerComponent implements OnInit {
 	annotationLayerChecked;
 
-	public LayersContainers$: Observable<any> = this.store.select(selectLayersContainers)
+	public layers$: Observable<any> = this.store.select(selectLayers)
 		.distinctUntilChanged()
-		.map((layersContainer: LayersContainer[]) => {
+		.map((layersContainer: Layer[]) => {
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts]: s/layerContainer/layer

File: src/app/@ansyn/menu-items/layers-manager/components/layers-collection/layer-collection.component.ts [hunk 11]
```diff
@@ -33,18 +40,49 @@ import { SelectLayerAction, UnselectLayerAction } from '@ansyn/menu-items/layers
...
-	public onCheckboxClicked(event, layer: Layer): void {
+	public onCheckboxClicked(event, layersContainer: Layer): void {
 		if (event.target.checked) {
-			this.store.dispatch(new SelectLayerAction(layer));
+			this.store.dispatch(new SelectLayerAction(layersContainer));
 		} else {
-			this.store.dispatch(new UnselectLayerAction(layer));
+			this.store.dispatch(new UnselectLayerAction(layersContainer));
 		}
+		this.collectionWithIsChecked = this.collectionWithIsChecked.map((layer) => {
```
Comment (liorze) [inline src/app/@ansyn/menu-items/layers-manager/components/layers-collection/layer-collection.component.ts]: I think a better approach to manage the selected layers would be using a Set. You don't need another "withIsChecked" collection. Tick a checkbox of layer.id is in the set, remove/add an id from/to the set onCheckboxClicked and simply pass the set to the reducer.

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.spec.ts [hunk 12]
```diff
@@ -7,7 +7,7 @@ import { SelectCaseAppEffects } from '@ansyn/ansyn/app-effects/effects/cases/sel
 import { cold, hot } from 'jasmine-marbles';
 import {
 	BeginLayerCollectionLoadAction,
-	ToggleDisplayAnnotationsLayer
+	ToggleDisplayAnnotationsLayer, UpdateSelectedLayersFromCaseAction
```
Comment (yairtawil) [inline src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.spec.ts]: rename `UpdateSelectedLayersFromCaseAction` to `UpdateSelectedLayersIds`

File: src/app/@ansyn/ansyn/app-effects/effects/layers.app.effects.ts [hunk 13]
```diff
@@ -1,29 +1,10 @@
...
-	 */
-	@Effect()
-	redrawMaps$: Observable<any> = this.actions$
-		.ofType<any>(LayersActionTypes.SELECT_LAYER, LayersActionTypes.UNSELECT_LAYER)
-		.map(() => new ContainerChangedTriggerAction());
 
-	constructor(protected actions$: Actions,
-				protected store$: Store<IAppState>,
-				protected imageryCommunicatorService: ImageryCommunicatorService) {
+	constructor() {
```
Comment (yairtawil) [inline src/app/@ansyn/ansyn/app-effects/effects/layers.app.effects.ts]: ❌ constructor

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 14]
```diff
@@ -57,9 +52,12 @@ import { ImageryProviderService } from '@ansyn/imagery/provider-service/imagery-
 import { ImageryCommunicatorService } from '@ansyn/imagery/communicator-service/communicator.service';
 import { MapFacadeService } from '@ansyn/map-facade/services/map-facade.service';
 import { CommunicatorEntity } from '@ansyn/imagery/communicator-service/communicator.entity';
-import { map, mergeMap } from 'rxjs/operators';
+import { map, mergeMap, tap } from 'rxjs/operators';
```
Comment (yairtawil) [inline src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts]: 💥

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 15]
```diff
@@ -88,15 +86,15 @@ export class MapAppEffects {
...
-			.map(({ payload}: DisplayOverlayAction) => new SetIsLoadingAcion({
-				mapId: payload.mapId, show: false
-			}));
+	onDisplayOverlayHideLoader$ = this.actions$
+		.ofType<DisplayOverlayAction>(
+			OverlaysActionTypes.DISPLAY_OVERLAY_SUCCESS,
+			OverlaysActionTypes.DISPLAY_OVERLAY_FAILED,
+			CoreActionTypes.BACK_TO_WORLD_VIEW
+		)
+		.map(({ payload }: DisplayOverlayAction) => new SetIsLoadingAcion({
```
Comment (yairtawil) [inline src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts]: pipe it

File: src/app/@ansyn/core/models/case.model.ts [hunk 16]
```diff
@@ -114,7 +114,8 @@ export interface CaseFacetsState {
 
 export interface CaseLayersState {
 	annotationsLayer: FeatureCollection<any>,
-	displayAnnotationsLayer?: boolean
+	displayAnnotationsLayer?: boolean,
+	activeLayersIds?: string[]
```
Comment (yairtawil) [inline src/app/@ansyn/core/models/case.model.ts]: ?

File: src/app/@ansyn/menu-items/cases/reducers/cases.reducer.ts [hunk 17]
```diff
@@ -70,3 +70,4 @@ export const selectCaseTotal = createSelector(casesStateSelector, selectTotal);
 export const selectCaseEntities = <MemoizedSelector<ICasesState, Dictionary<CasePreview>>>createSelector(casesStateSelector, selectEntities);
 export const selectCasesIds = <MemoizedSelector<any, string[] | number[]>>createSelector(casesStateSelector, selectIds);
 export const selectSelectedCase = createSelector(casesStateSelector, (cases) => cases && cases.selectedCase);
+export const selectedLayersIds = createSelector(casesStateSelector , (cases) => cases && cases.selectedCase && cases.selectedCase.state.layers.activeLayersIds );
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/cases/reducers/cases.reducer.ts]: removeeeeee

move `selectedLayersIds` to layers.reducer

File: src/app/@ansyn/menu-items/layers-manager/actions/layers.actions.ts [hunk 18]
```diff
@@ -1,25 +1,23 @@
...
 export const LayersActionTypes = {
 	BEGIN_LAYER_COLLECTION_LOAD: 'BEGIN_LAYER_COLLECTION_LOAD',
 	LAYER_COLLECTION_LOADED: 'LAYER_COLLECTION_LOADED',
-	SELECT_LAYER: 'SELECT_LAYER',
-	UNSELECT_LAYER: 'UNSELECT_LAYER',
 	ERROR_LOADING_LAYERS: 'ERROR_LOADING_LAYERS',
 	ANNOTATIONS: {
 		TOGGLE_DISPLAY_LAYER: 'TOGGLE_LAYER'
-	}
+	},
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/actions/layers.actions.ts]: remove one of them

File: src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts [hunk 19]
```diff
@@ -17,11 +17,11 @@ import { groupBy } from 'lodash';
 export class LayersManagerComponent implements OnInit {
 	annotationLayerChecked;
 
-	public LayersContainers$: Observable<any> = this.store.select(selectLayersContainers)
+	public layers$: Observable<any> = this.store.select(selectLayers)
 		.distinctUntilChanged()
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts]: redundent `distinctUntilChanged`

File: src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts [hunk 20]
```diff
@@ -33,7 +33,7 @@ export class LayersManagerComponent implements OnInit {
 			.subscribe(result => {
 				this.annotationLayerChecked = result;
 			});
-		this.LayersContainers$.subscribe();
+		this.layers$.subscribe();
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/components/layers-manager/layers-manager.component.ts]: move to ngOnInit

File: src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.ts [hunk 21]
```diff
@@ -1,10 +1,5 @@
 import { ILayerState } from '../reducers/layers.reducer';
-import {
-	ErrorLoadingLayersAction,
-	LayerCollectionLoadedAction,
-	LayersActions,
-	LayersActionTypes
-} from '../actions/layers.actions';
+import { LayerCollectionLoadedAction, LayersActions, LayersActionTypes } from '../actions/layers.actions';
 import 'rxjs/add/operator/map';
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/effects/layers.effects.ts]: pipe it ? please

File: src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.ts [hunk 22]
```diff
@@ -22,19 +25,15 @@ export function LayersReducer(state: ILayerState = initialLayersState, action: L
...
-
-		case LayersActionTypes.UNSELECT_LAYER:
-			action.payload.isChecked = false;
-			return { ...state, layersContainers: [ ...state.layersContainers ] };
+			return { ...state, layers: action.payload };
 
 		case LayersActionTypes.ANNOTATIONS.TOGGLE_DISPLAY_LAYER:
 			return { ...state, displayAnnotationsLayer: action.payload };
 
+		case LayersActionTypes.UPDATE_SELECTED_LAYERS_FROM_CASE:
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.ts]: omg

File: src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.ts [hunk 23]
```diff
@@ -44,5 +43,6 @@ export function LayersReducer(state: ILayerState = initialLayersState, action: L
 
 }
 
-export const selectLayersContainers = createSelector(layersStateSelector, (layersState: ILayerState) => layersState.layersContainers);
-export const selectDisplayAnnotationsLayer = createSelector(layersStateSelector, (layersState: ILayerState) => layersState ? layersState.displayAnnotationsLayer : true);
+export const selectLayers = createSelector(layersStateSelector, (layersState: ILayerState) => layersState.layers);
+export const selectDisplayAnnotationsLayer = createSelector(layersStateSelector, (layersState: ILayerState) => layersState.displayAnnotationsLayer);
+export const selectSelectedLayersIds = createSelector(layersStateSelector, (layersState: ILayerState) => layersState.selectedLayersIds);
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/reducers/layers.reducer.ts]: 👋

File: src/app/@ansyn/menu-items/layers-manager/services/data-layers.service.ts [hunk 24]
```diff
@@ -5,7 +5,7 @@ import 'rxjs/add/operator/catch';
 import 'rxjs/add/operator/map';
 import { StorageService } from '@ansyn/core/services/storage/storage.service';
 import { ErrorHandlerService } from '@ansyn/core/services/error-handler.service';
-import { LayersContainer } from '@ansyn/menu-items/layers-manager/models/layers.model';
+import { Layer } from '@ansyn/menu-items/layers-manager/models/layers.model';
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/layers-manager/services/data-layers.service.ts]: try to replace to `ILayer`

**Pull request merged**